ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V1/Common.hs

```diff
@@ -0,0 +1,256 @@
+import           Data.Word
+import           GHC.Generics (Generic)
+import           NoThunks.Class
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.SupportsProtocol
+import           Ouroboros.Consensus.Storage.LedgerDB.API as API
+import           Ouroboros.Consensus.Storage.LedgerDB.API.Config
+import           Ouroboros.Consensus.Storage.LedgerDB.Impl.Common
+import           Ouroboros.Consensus.Storage.LedgerDB.Impl.Snapshots
+import           Ouroboros.Consensus.Storage.LedgerDB.Impl.Validate
+import           Ouroboros.Consensus.Storage.LedgerDB.V1.Args
+import           Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore
+import           Ouroboros.Consensus.Storage.LedgerDB.V1.DbChangelog
+import           Ouroboros.Consensus.Storage.LedgerDB.V1.Lock
+import           Ouroboros.Consensus.Util.CallStack
+import           Ouroboros.Consensus.Util.IOLike
+
+{-------------------------------------------------------------------------------
+  LedgerDB internal state
+-------------------------------------------------------------------------------}
+
+newtype LedgerDBHandle m l blk = LDBHandle (StrictTVar m (LedgerDBState m l blk))
+  deriving Generic
+
+data LedgerDBState m l blk =
+    LedgerDBOpen !(LedgerDBEnv m l blk)
+  | LedgerDBClosed
+  deriving Generic
+
+deriving instance ( IOLike m
+                  , LedgerSupportsProtocol blk
+                  , NoThunks (l EmptyMK)
+                  , NoThunks (Key l)
+                  , NoThunks (Value l)
+                  , NoThunks (LedgerCfg l)
+                  ) => NoThunks (LedgerDBState m l blk)
+
+type LedgerDBEnv :: (Type -> Type) -> LedgerStateKind -> Type -> Type
+data LedgerDBEnv m l blk = LedgerDBEnv {
+    -- | INVARIANT: the tip of the 'LedgerDB' is always in sync with the tip of
+    -- the current chain of the ChainDB.
+    ldbChangelog      :: !(StrictTVar m (DbChangelog l))
+    -- | Handle to the ledger's backing store, containing the parts that grow too
+    -- big for in-memory residency
+  , ldbBackingStore   :: !(LedgerBackingStore m l)
+    -- | The flush lock to the 'BackingStore'. This lock is crucial when it
+    -- comes to keeping the data in memory consistent with the data on-disk.
+    --
+    -- This lock should be held whenever we want to keep a consistent view of
+    -- the backing store for some time. In particular we use this:
+    --
+    -- - when performing a query on the ledger state, we need to hold a
+    --   'LocalStateQueryView' which, while live, must maintain a consistent view
+    --   of the DB, and therefore we acquire a Read lock.
+    --
+    -- - when taking a snapshot of the ledger db, we need to prevent others
+    --   from altering the backing store at the same time, thus we acquire a
+    --   Write lock.
+  , ldbLock           :: !(AllowThunk (LedgerDBLock m))
```

Comment:
Short answer: No we can't at the moment.

Long answer: We need something like this for TMVars:
```
instance NoThunks a => NoThunks (StrictSTM.StrictTVar IO a) where
  showTypeOf _ = "StrictTVar IO"
  wNoThunks ctxt tv = do
      -- We can't use @atomically $ readTVar ..@ here, as that will lead to a
      -- "Control.Concurrent.STM.atomically was nested" exception.
      a <- StrictSTM.readTVarIO tv
      noThunks ctxt a
```
because of the same issue with not being able to atomically read the tmvar.